publication-commons/src/main/java/no/unit/nva/doi/models/ImmutableDoi.java

```diff
@@ -0,0 +1,369 @@
+            this.suffix = Objects.requireNonNull(suffix, MESSAGE_NON_NULL_ARGUMENT_FOR_PARAMETER_SUFFIX);
+            initBits &= ~INIT_BIT_SUFFIX;
+            return this;
+        }
+
+        /**
+         * Initializes the value for the {@link Doi#getPrefix()} and {@link Doi#getSuffix()} attributes.
+         *
+         * @param identifier The value (doi identifier: prefix/suffix) that can be parsed into prefix and suffix.
+         * @return {@code this} builder for use in a chained invocation
+         */
+        public final Builder withIdentifier(String identifier) {
+            Objects.requireNonNull(identifier, MESSAGE_NON_NULL_ARGUMENT_FOR_PARAMETER_IDENTIFIER);
+            int indexOfDivider = identifier.indexOf(FORWARD_SLASH);
+            if (indexOfDivider == -1) {
+                throw new IllegalArgumentException("Invalid DOI identifier");
+            }
+            withPrefix(identifier.substring(0, indexOfDivider));
+            withSuffix(identifier.substring(++indexOfDivider));
+            return this;
+        }
+
+        /**
+         * Initializes the value for {@link Doi#getProxy()}, {@link Doi#getPrefix()} and {@link Doi#getSuffix()}
+         * attributes.
+         *
+         * @param doi Teh value (Doi Id: https://proxy/prefix/suffix) that can be parsed into proxy, prefix and suffix.
+         * @return {@code this} builder for use in a chained invocation
+         */
+        public final Builder withDoi(URI doi) {
+            Objects.requireNonNull(doi, MESSAGE_NON_NULL_ARGUMENT_FOR_PARAMETER_DOI);
+            if (containsOnlyLeadingForwardSlashAndSlashBetweenPrefixAndSuffix(doi)) {
+                throw new IllegalArgumentException(ERROR_DOI_URI_INVALID_FORMAT.concat(doi.toASCIIString()));
+            }
+            withIdentifier(extractDoiPathWithoutLeadingForwardSlash(doi));
+            withProxy(doi);
+            return this;
+        }
+
+        /**
+         * Builds a new {@link ImmutableDoi ImmutableDoi}.
+         *
+         * @return An immutable instance of Doi
+         * @throws java.lang.IllegalStateException if any required attributes are missing
+         */
+        public ImmutableDoi build() {
+            checkRequiredAttributes();
+            return new ImmutableDoi(this);
+        }
+
+        private static String extractDoiPathWithoutLeadingForwardSlash(URI doi) {
+            return doi.getPath().charAt(0) == FORWARD_SLASH ? doi.getPath().substring(1) : doi.getPath();
+        }
+
+        private static boolean containsOnlyLeadingForwardSlashAndSlashBetweenPrefixAndSuffix(URI doi) {
+            return doi.getPath().chars().filter(ch -> ch == FORWARD_SLASH).count() != 2;
+        }
+
+        private static void checkNotIsSet(boolean isSet, String name) {
+            if (isSet) {
+                throw new IllegalStateException("Builder of Doi is strict, attribute is already set: ".concat(name));
+            }
+        }
+
+        private boolean proxyIsSet() {
```

Comment:
These in essence replace nonNull(field), do they not?

Reply:
So the base is based on org.immutables which we have inhoused.
I trust the checks are faster from the generated code I used as a base.

Reply:
Why is it necessary to check in the builder if something has been set already? Why to make the builder "immutable" also?

Reply:
For avoid calling same methods twice and validating proxy as proxy argument is a URI when we only use scheme, authority and port. Also note proxy has a default argument if not provided and then is not set in the builder.